crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,252 @@
+                            }
+
+                            if finished.len() >= task_limit {
+                                break;
+                            }
+                        }
+                    }
+                    // In this case, messages in the mailbox have been overwritten
+                    // as a result of lagging too far behind. We can simply continue the loop,
+                    // as the next call to recv() should return the overwritten checkpoint and we
+                    // will attempt to execute all checkpoints from (highest_executed_checkpoint + 1)
+                    // up to the result of the next recv() call.
+                    Err(RecvError::Lagged(num_skipped)) => {
+                        warn!(
+                            "Checkpoint Execution Recv channel overflowed {:?} messages",
+                            num_skipped,
+                        );
+                        self.metrics
+                            .checkpoint_exec_recv_channel_overflow
+                            .inc_by(num_skipped);
+                    }
+                    Err(RecvError::Closed) => {
+                        info!("Checkpoint Execution Sender (StateSync) closed channel");
+                        break;
+                    }
+                },
+            }
+        }
+    }
+
+    async fn reconfig(&self, _next_epoch_committee: &[(AuthorityName, StakeUnit)]) -> SuiResult {
+        // TODO(william) call into Reconfig
+        Ok(())
+    }
+
+    fn update_lag_metric(&self, highest_executed: CheckpointSequenceNumber) -> SuiResult {
+        let highest_synced = self
+            .checkpoint_store
+            .get_highest_synced_checkpoint_seq_number()?
+            .unwrap_or(0);
+
+        let diff = highest_synced - highest_executed;
+        self.metrics.checkpoint_exec_lag.set(diff as i64);
+        Ok(())
+    }
+}
+
+pub async fn execute_checkpoint(
+    checkpoint: VerifiedCheckpoint,
+    authority_state: Arc<AuthorityState>,
+    checkpoint_store: Arc<CheckpointStore>,
+) -> SuiResult<Vec<TransactionEffects>> {
+    let txes = checkpoint_store
+        .get_checkpoint_contents(&checkpoint.content_digest())
+        .map_err(SuiError::from)?
+        .unwrap_or_else(|| {
+            panic!(
+                "Checkpoint contents for digest {:?} does not exist",
+                checkpoint.content_digest()
+            )
+        })
+        .into_inner();
+
+    let effects = execute_and_verify_transactions(txes, authority_state).await?;
+    Ok(effects)
+}
+
+async fn execute_and_verify_transactions(
+    transactions: Vec<ExecutionDigests>,
+    authority_state: Arc<AuthorityState>,
+) -> SuiResult<Vec<TransactionEffects>> {
+    let tx_digests: Vec<TransactionDigest> = transactions.iter().map(|tx| tx.transaction).collect();
+    let txns = authority_state
+        .database
+        .multi_get_pending_certificate(&tx_digests)?
```

Comment:
Since state sync process is responsible for getting all the certs, I wonder why not have state sync just pass the vector of certs directly into executor, instead of reading again here? Mostly worrying about that the per-epoch table may get wiped out if we are unlucky and a reconfiguration happened

Reply:
There's a chance that we overflow the channel between state sync and executor, in which case messages would be overwritten. So either way we will have the need to recover tx's and checkpoint certs from storage.